core/include/detray/definitions/detail/accessor.hpp

```diff
@@ -95,5 +101,67 @@ DETRAY_HOST_DEVICE constexpr tuple_type<unwrap_decay_t<Types>...> make_tuple(
     return tuple_type<unwrap_decay_t<Types>...>(std::forward<Types>(args)...);
 }
 
+/**
+ *  Trait class to figure out if a given type has a @c reserve(...) function
+ */
+template <typename T>
+struct has_reserve {
+
+    private:
+    /// Function returning @c std::true_type for types that do have a @c
+    /// reserve(...) function
+    template <typename C>
+    static constexpr auto check(C*) ->
+        typename std::is_void<decltype(std::declval<C>().reserve(
+            std::declval<typename C::size_type>()))>::type;
+    /// Function returning @c std::false_type for types that fair the previous
+    /// function
+    template <typename>
+    static constexpr std::false_type check(...);
+
+    /// Declare the value type of this trait class
+    typedef decltype(check<T>(nullptr)) type;
+
+    public:
+    /// Value of the check
+    static constexpr bool value = type::value;
+};
+
+/// @name Functions calling or not calling reserve(...) based on whether it's
+/// available
+/// @{
+
+template <typename T, std::enable_if_t<has_reserve<T>::value, bool> = true>
+DETRAY_HOST_DEVICE void call_reserve(T& obj, std::size_t newsize) {
+    obj.reserve(newsize);
+}
+
+template <typename T, std::enable_if_t<!(has_reserve<T>::value), bool> = true>
+DETRAY_HOST_DEVICE void call_reserve(T& obj, std::size_t newsize) {}
```

Comment:
We don't have to necessarily do it in this PR, but...

I was thinking... for the case where the vector type doesn't have a `reserve()` function, we should probably at least add the following to this function:

```c++
assert(obj.capacity() >= newsize);
```

Just so that if in device code we don't give a "big enough" container to the object, that would be a bit easier to debug.

We can probably skip introducing a `has_capacity` type though, as we should really never call `call_reserve(...)` on a type that doesn't even have that function.

So, since the change should really just be to add one statement here, I would vote for doing it in this PR already. And note that we would probably need to do it like:

```c++
template <typename T, std::enable_if_t<!(has_reserve<T>::value), bool> = true>
DETRAY_HOST_DEVICE void call_reserve(T&
#ifndef NDEBUG
    obj
#endif  // not NDEBUG
    , std::size_t
#ifndef NDEBUG
    newsize
#endif  // not NDEBUG
    ) { assert(obj.capacity() >= newsize); }
```

Yeah, I hate these things too. Where you need to hide the name of the variables from the compiler in optimised mode, as in that mode they are not used anywhere in the function body...